akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.http.impl.engine.ws
+
+import scala.concurrent.Await
+import scala.concurrent.duration.DurationInt
+import org.scalactic.ConversionCheckedTripleEquals
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.time.Span.convertDurationToSpan
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.HttpRequest
+import akka.http.scaladsl.model.Uri.apply
+import akka.http.scaladsl.model.ws._
+import akka.stream._
+import akka.stream.scaladsl._
+import akka.stream.testkit._
+import akka.stream.scaladsl.GraphDSL.Implicits._
+import org.scalatest.concurrent.Eventually
+import akka.stream.io.SslTlsPlacebo
+import java.net.InetSocketAddress
+import akka.stream.impl.fusing.GraphStages
+import akka.util.ByteString
+
+class WebsocketIntegrationSpec extends AkkaSpec("akka.stream.materializer.debug.fuzzing-mode=off")
+  with ScalaFutures with ConversionCheckedTripleEquals with Eventually {
+
+  implicit val patience = PatienceConfig(3.seconds)
+  import system.dispatcher
+  implicit val materializer = ActorMaterializer()
+
+  "A Websocket server" must {
+
+    "echo 100 elements and then shut down without error" in Utils.assertAllStagesStopped {
+      val bindingFuture = Http().bindAndHandleSync({
+        case HttpRequest(_, _, headers, _, _) ⇒
+          val upgrade = headers.collectFirst { case u: UpgradeToWebsocket ⇒ u }.get
+          upgrade.handleMessages(Flow.apply, None)
+      }, interface = "localhost", port = 8080)
+      val binding = Await.result(bindingFuture, 3.seconds)
+
+      val N = 100
+      val (response, count) = Http().singleWebsocketRequest(
+        WebsocketRequest("ws://127.0.0.1:8080"),
+        Flow.fromSinkAndSourceMat(
+          Sink.fold(0)((n, _: Message) ⇒ n + 1),
+          Source.repeat(TextMessage("hello")).take(N))(Keep.left))
+
+      count.futureValue should ===(N)
+      binding.unbind()
+    }
+
+    "send back 100 elements and then terminate without error even when not ordinarily closed" in Utils.assertAllStagesStopped {
+      val N = 100
```

Comment:
detail: could move N out of test and use in test message `s"send back #N elements..."`